plugins/webp-uploads/hooks.php

```diff
@@ -681,38 +681,54 @@ function webp_uploads_update_featured_image( $html, $post_id, $attachment_id ) {
-	$preserve_jpegs_for_webp_transforms = isset( $transforms['image/webp'] ) && in_array( 'image/jpeg', $transforms['image/webp'], true ) && in_array( 'image/webp', $transforms['image/webp'], true );
-	if ( ! $preserve_jpegs_for_jpeg_transforms && ! $preserve_jpegs_for_webp_transforms ) {
+	$transforms   = webp_uploads_get_upload_image_mime_transforms();
+	$image_format = get_option( 'perflab_modern_image_format' );
+	$image_format = empty( $image_format ) ? 'webp' : $image_format;
+
+	// We need to add fallback only if jpeg alternatives for the image_format images are enabled for the server.
+	$preserve_jpegs_for_jpeg_transforms       = isset( $transforms['image/jpeg'] ) && in_array( 'image/jpeg', $transforms['image/jpeg'], true ) && in_array( 'image/' . $image_format, $transforms['image/jpeg'], true );
+	$preserve_jpegs_for_image_type_transforms = isset( $transforms[ 'image/' . $image_format ] ) && in_array( 'image/jpeg', $transforms[ 'image/' . $image_format ], true ) && in_array( 'image/' . $image_format, $transforms[ 'image/' . $image_format ], true );
+	if ( ! $preserve_jpegs_for_jpeg_transforms && ! $preserve_jpegs_for_image_type_transforms ) {
 		return;
 	}
 
-	ob_start();
+	$detection_string = '';
+	// The fallback script can only handle a single image format at a time.
+	if ( 'webp' === $image_format ) {
+		$detection_string = 'data:image/webp;base64,UklGRh4AAABXRUJQVlA4TBEAAAAvAQAAAAfQ//73v/+BiOh/AAA=';
+	} elseif ( 'avif' === $image_format ) {
+		$detection_string = 'data:image/avif;base64,AAAAIGZ0eXBhdmlmAAAAAGF2aWZtaWYxbWlhZk1BMUIAAADybWV0YQAAAAAAAAAoaGRscgAAAAAAAAAAcGljdAAAAAAAAAAAAAAAAGxpYmF2aWYAAAAADnBpdG0AAAAAAAEAAAAeaWxvYwAAAABEAAABAAEAAAABAAABGgAAAB0AAAAoaWluZgAAAAAAAQAAABppbmZlAgAAAAABAABhdjAxQ29sb3IAAAAAamlwcnAAAABLaXBjbwAAABRpc3BlAAAAAAAAAAIAAAACAAAAEHBpeGkAAAAAAwgICAAAAAxhdjFDgQ0MAAAAABNjb2xybmNseAACAAIAAYAAAAAXaXBtYQAAAAAAAAABAAEEAQKDBAAAACVtZGF0EgAKCBgANogQEAwgMg8f8D///8WfhwB8+ErK42A=';
+	}
+
+	if ( empty( $detection_string ) ) {
+		return;
+	}
 
-	?>
-	( function( d, i, s, p ) {
+	$js_function = <<<JS
+	function webpUploadsDetectFallback( d, i, s, p, fallbackSrc ) {
 		s = d.createElement( s );
-		s.src = '<?php echo esc_url_raw( plugins_url( '/fallback.js', __FILE__ ) ); ?>';
+		s.src = fallbackSrc;
 
 		i = d.createElement( i );
-		i.src = p + 'jIAAABXRUJQVlA4ICYAAACyAgCdASoCAAEALmk0mk0iIiIiIgBoSygABc6zbAAA/v56QAAAAA==';
+		i.src = p;
 		i.onload = function() {
 			i.onload = undefined;
-			i.src = p + 'h4AAABXRUJQVlA4TBEAAAAvAQAAAAfQ//73v/+BiOh/AAA=';
+			i.src = 'data:image/webp;base64,UklGRh4AAABXRUJQVlA4TBEAAAAvAQAAAAfQ//73v/+BiOh/AAA=';
 		};
 
 		i.onerror = function() {
 			d.body.appendChild( s );
 		};
-	} )( document, 'img', 'script', 'data:image/webp;base64,UklGR' );
-	<?php
-	$javascript = ob_get_clean();
+	}
+JS;
 
 	wp_print_inline_script_tag(
-		preg_replace( '/\s+/', '', $javascript ),
+		sprintf(
+			'( %s )( "document", "img", "script", %s, %s )',
+			preg_replace( '/\s+/', '', $js_function ),
```

Comment:
I'm surprised the JS isn't erroring with removal of all whitespace. I think this would be safer. although it doesn't cause an error currently:

```suggestion
			preg_replace( '/\s+/', ' ', $js_function ),
```

Reply:
why, isn't whitespace ignored in JavaScript?

Reply:
I was thinking of JS like `var foo = bar`, if the whitespace is removed then this becomes `varfoo=bar` which would be an error. This isn't an issue with the JS as written right now though.

Reply:
oh good point, thanks for clarifying. accepting the suggestion!